File: lessons/00-workflow-or-agent/typescript/src/schemas.ts
import {readFileSync} from "node:fs";

import Ajv2020, {type ErrorObject, type ValidateFunction} from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

import {EVENT_SCHEMA_PATH, RESULT_SCHEMA_PATH, ROUTE_SCHEMA_PATH} from "./paths.js";
import type {RouteDecision, RunEvent, RunResult, ValidationResult} from "./types.js";

function loadSchema(path: string): object {
  return JSON.parse(readFileSync(path, "utf8")) as object;
}

const ajv = new Ajv2020({allErrors: true, strict: true});
addFormats(ajv);

const routeValidator = ajv.compile<RouteDecision>(loadSchema(ROUTE_SCHEMA_PATH));
const eventValidator = ajv.compile<RunEvent>(loadSchema(EVENT_SCHEMA_PATH));
const resultValidator = ajv.compile<RunResult>(loadSchema(RESULT_SCHEMA_PATH));

function stableErrors(errors: ErrorObject[] | null | undefined): string[] {
  return (errors ?? []).map((error) => {
    const path = error.instancePath === "" || error.instancePath === undefined ? "/" : error.instancePath;
    return `${path} ${error.message ?? "is invalid"}`;
  });
}

function validate<T>(validator: ValidateFunction<T>, raw: unknown): ValidationResult<T> {
  if (validator(raw)) {
    return {ok: true, value: structuredClone(raw)};
  }
  return {ok: false, errors: stableErrors(validator.errors)};
}

export function validateRouteDecision(raw: unknown): ValidationResult<RouteDecision> {
  return validate(routeValidator, raw);
}

export function validateRunEvent(raw: unknown): ValidationResult<RunEvent> {
  return validate(eventValidator, raw);
}

export function validateRunResult(raw: unknown): ValidationResult<RunResult> {
  return validate(resultValidator, raw);
}

File: lessons/00-workflow-or-agent/python/src/boundrelay_m0/schemas.py
from copy import deepcopy
from dataclasses import dataclass
import json
import math
from typing import Generic, Literal, Mapping, TypeVar, cast

from jsonschema import Draft202012Validator, FormatChecker

from .paths import EVENT_SCHEMA_PATH, RESULT_SCHEMA_PATH, ROUTE_SCHEMA_PATH
from .types import Route, RouteDecision

T = TypeVar("T")


@dataclass(frozen=True)
class ValidationSuccess(Generic[T]):
    value: T

    @property
    def ok(self) -> Literal[True]:
        return True


@dataclass(frozen=True)
class ValidationFailure:
    errors: tuple[str, ...]

    @property
    def ok(self) -> Literal[False]:
        return False


ValidationResult = ValidationSuccess[T] | ValidationFailure


def _load_schema(path) -> dict[str, object]:
    value = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(value, dict):
        raise ValueError(f"Schema must be an object: {path}")
    return value


_FORMAT_CHECKER = FormatChecker()
_ROUTE_VALIDATOR = Draft202012Validator(_load_schema(ROUTE_SCHEMA_PATH), format_checker=_FORMAT_CHECKER)
_EVENT_VALIDATOR = Draft202012Validator(_load_schema(EVENT_SCHEMA_PATH), format_checker=_FORMAT_CHECKER)
_RESULT_VALIDATOR = Draft202012Validator(_load_schema(RESULT_SCHEMA_PATH), format_checker=_FORMAT_CHECKER)


def _stable_errors(validator: Draft202012Validator, raw: object) -> tuple[str, ...]:
    errors = sorted(validator.iter_errors(raw), key=lambda error: (list(error.absolute_path), error.message))
    values: list[str] = []
    for error in errors:
        path = "/" + "/".join(str(part) for part in error.absolute_path)
        values.append(f"{path or '/'} {error.message}")
    return tuple(values)


def _validate_mapping(
    validator: Draft202012Validator,
    raw: object,
) -> ValidationSuccess[dict[str, object]] | ValidationFailure:
    errors = _stable_errors(validator, raw)
    if errors:
        return ValidationFailure(errors)
    if not isinstance(raw, Mapping):
        return ValidationFailure(("/ must be an object",))
    return ValidationSuccess(deepcopy(dict(raw)))


def validate_route_decision(raw: object) -> ValidationSuccess[RouteDecision] | ValidationFailure:
    if isinstance(raw, Mapping):
        confidence = raw.get("confidence")
        if isinstance(confidence, int) and not isinstance(confidence, bool):
            if confidence < 0 or confidence > 1:
                return ValidationFailure(("/confidence must be between 0 and 1",))
        elif isinstance(confidence, float) and not math.isfinite(confidence):
            return ValidationFailure(("/confidence must be a finite number",))

    validated = _validate_mapping(_ROUTE_VALIDATOR, raw)
    if not validated.ok:
        return validated
    value = validated.value
    return ValidationSuccess(RouteDecision(cast(Route, value["route"]), float(value["confidence"])))


def validate_run_event(raw: object) -> ValidationSuccess[dict[str, object]] | ValidationFailure:
    return _validate_mapping(_EVENT_VALIDATOR, raw)


def validate_run_result(raw: object) -> ValidationSuccess[dict[str, object]] | ValidationFailure:
    return _validate_mapping(_RESULT_VALIDATOR, raw)
